pkg/traces/servicegraphprocessor/processor.go

```diff
@@ -21,36 +18,61 @@ import (
 	"google.golang.org/grpc/codes"
 )
 
-var (
-	errTooManyItems = errors.New("too many items in store")
-)
+type tooManySpansError struct {
+	droppedSpans int
+}
+
+func (t tooManySpansError) Error() string {
+	return fmt.Sprintf("dropped %d spans", t.droppedSpans)
+}
+
+// edge is an edge between two nodes in the graph
+type edge struct {
+	key string
 
-// edgeRequest is a request between two nodes in the graph
-type edgeRequest struct {
 	serverService, clientService string
 	serverLatency, clientLatency time.Duration
 
 	// If either the client or the server spans have status code error,
-	// the request will be considered as failed.
+	// the edge will be considered as failed.
 	failed bool
+
+	// expiration is the time at which the edge expires, expressed as Unix time
+	expiration int64
 }
 
-// complete returns true if the corresponding client and server
-// pair spans have been processed for the given request
-func (e *edgeRequest) complete() bool {
+func newEdge(key string, ttl time.Duration) *edge {
+	return &edge{
+		key: key,
+
+		expiration: time.Now().Add(ttl).Unix(),
+	}
+}
+
+// completed returns true if the corresponding client and server
+// pair spans have been processed for the given edge
+func (e *edge) isCompleted() bool {
```

Comment:
`isCompleted()` has a weird edge case where if a span has a `service.name` attribute but it's an empty string then this function will never return completed even if both sides of an edge has been discovered.

When we get the service name perhaps we could do something like:

```
svc, ok := rSpan.Resource().Attributes().Get(semconv.AttributeServiceName)
if !ok {
	continue
}
// substitute in a service name ....
if len(svc) == 0 {
  svc = "<empty>"
}
```

Reply:
You're right, nice catch. Although I'm not convinced that we should process spans with empty service name. I'd rather count them as unpaired.

Reply:
Pushed a change to skip processing of spans with `""` service name. It can be changed later.

Reply:
I'm good on this, but make sure that the documentation lists `service.name` (along with `span.kind`) as a required attributes for this to work.